### PR TITLE
Better rust_link handling; add module-based allowlist

### PR DESCRIPTION
**Pull request by Manishearth**
Blocked on https://github.com/rust-diplomat/diplomat/pull/224

Can't actually be run until https://github.com/unicode-org/icu4x/pull/2403

This updates to a Diplomat with more `rust_link` options; but does not yet use them. I'll do a pass once the rest of this merges.

It also adds an initial allowlist, split into two sections: one section for things that don't need to be exposed over FFI (because they are rust-specific) and one for things which coudl be but aren't currently planned.

The hope is that the latter section of the allowlist can be pared down over time. The nice thing is that by keeping it in source rather than completely in `missing_keys` we can comment it, and hopefulyl we can get to a point where `missing_keys` is empty.

### CODE REVIEW

**Comment by Manishearth** [comment]
Once this and #2403 land I plan to do a pass through of our current FFI and ensure everything is maximally tagged with rustlinks.

**Comment by Manishearth** [comment]
cc @robertbastian

File: ffi/diplomat/src/bin/completeness.rs [hunk 1]
```diff
@@ -91,6 +108,45 @@ lazy_static::lazy_static! {
         "Yokeable",
         "ZeroFrom",
     ].into_iter().collect();
+
+    static ref ALLOWLISTED_PATHS: HashSet<Vec<String>> = [
```
Comment (robertbastian) [inline ffi/diplomat/src/bin/completeness.rs:112]: I think "allow" is not a good phrase (are we "allowing" it to be excluded). I'd call it IGNORELIST

File: ffi/diplomat/src/bin/completeness.rs [hunk 2]
```diff
@@ -91,6 +108,45 @@ lazy_static::lazy_static! {
...
+        // Stuff that could be exposed over FFI but is not currently planned
+        // =========================
+        // Largely for use by datetimeformat, not super public
+        "icu::calendar::week_of",
```
Comment (sffc) [inline ffi/diplomat/src/bin/completeness.rs:116]: Nit: let's track this in https://github.com/unicode-org/icu4x/issues/2421

File: ffi/diplomat/src/bin/completeness.rs [hunk 3]
```diff
@@ -91,6 +108,45 @@ lazy_static::lazy_static! {
...
+
+
+        // Individual calendars: Currently the main entry point is AnyCalendar
+        "icu::calendar::iso",
```
Comment (sffc) [inline ffi/diplomat/src/bin/completeness.rs:120]: Nit: ISO should be covered (and I add it in my PR)

File: ffi/diplomat/src/bin/completeness.rs [hunk 4]
```diff
@@ -91,6 +108,45 @@ lazy_static::lazy_static! {
...
+        "icu::calendar::buddhist",
+        "icu::calendar::coptic",
+        "icu::calendar::ethiopian",
+        "icu::calendar::indian",
+        "icu::calendar::japanese",
+        "icu::calendar::julian",
+        "icu::calendar::any_calendar::IncludedInAnyCalendar",
```
Comment (sffc) [inline ffi/diplomat/src/bin/completeness.rs:127]: Nit: We renamed the trait

File: ffi/diplomat/src/bin/completeness.rs [hunk 5]
```diff
@@ -91,6 +108,45 @@ lazy_static::lazy_static! {
...
+        // =========================
+
+        // "Internal" trait that should never be called directly
+        "icu::calendar::Calendar",
+        // Used for rust-specific type transforms
+        "icu::calendar::AsCalendar",
+
+        // FFI largely deals with primitives rather than Rust's nice wrapper types
+        // (which are hard to do in a zero-cost way over FFI)
+        "icu::calendar::types",
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply ffi/diplomat/src/bin/completeness.rs:142]: I think _at least for 1.0_ we don't need them. And most languages don't do type wrapper safety the same way Rust does.

I'm imagining that this list _for now_ is a 1.0 list and in the future we change what it means: e.g. if we decide to support individual calendar types we can cut out the calendar chunk.

**Review by robertbastian — APPROVED** [review]

File: ffi/diplomat/src/bin/completeness.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Manishearth) [reply ffi/diplomat/src/bin/completeness.rs:143]: The main issue is that type wrappers aren't fully zero cost in e.g. JS, nor are they idiomatic.

[the remark this answers is not in the document]
Comment (sffc) [reply ffi/diplomat/src/bin/completeness.rs:143]: They mostly work in C++ though.

**Review by sffc — APPROVED** [review]

**Pull request merged**